### PR TITLE
feat: support async formatter

### LINKED ISSUE #15242: Change Request: Support async formatter
### ESLint version

8.1.0

### What problem do you want to solve?

Support async formatter

image

[fengzilong/eslint-formatter-mo](https://github.com/fengzilong/eslint-formatter-mo)

I had wrote an eslint formatter with code highlight feature

But it seems the highlighter didn't highlight all tokens well

Recently I want to switch highlighter to [shiki](https://github.com/shikijs/shiki) for a better highlighting result(maybe, I'm trying), but `shiki` has a async highlighter, which is not supported by ESLint currently

### What do you think is the correct solution?

`formatter` can be async function

### Participation

- [X] I am willing to submit a pull request for this change.

### Additional comments

_No response_

### PR DESCRIPTION
**Pull request by fengzilong**
closes #15242

#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[x] Add something to the core: support async formatter
[ ] Other, please explain

#### What changes did you make? (Give an overview)

Added `await` ahead of `formatter.format`

### CODE REVIEW

**Review by nzakas — CHANGES_REQUESTED** [review]
Thanks for this. We would just need some tests to verify this change works as intended.

We would also need the formatter documentation to be updated.

**Comment by fengzilong** [comment]
> We would just need some tests to verify this change works as intended.
> We would also need the formatter documentation to be updated.

I'll start working on this soon when I get off work

**Comment by fengzilong** [comment]
> We would just need some tests to verify this change works as intended.

Test was added

> We would also need the formatter documentation to be updated.

I'm not sure in which version will ESLint support async formatter. I think I'd better add documentation after the releasing, or it may mislead some formatter developers( I will add documentation then in another PR ). Hope to hear your thoughts

ping~ @nzakas

**Comment by nzakas** [comment]
Please add the docs in this PR. We release features and docs at the same time.

**Comment by fengzilong** [comment]
> Please add the docs in this PR. We release features and docs at the same time.

Docs added, and I used `from ESLint 8.2.0` in the documentation, not sure whether it is the right version number for next releasing, please help review this, thanks~

**Comment by fengzilong** [comment]
Some help is needed, I'd like to modify the `Formatter` type, but I just couldn't find where it is from, the `./load-formatter` is missing in codebase

https://github.com/eslint/eslint/blob/a1f7ad77e2da00ac7d6daade547fe6bef4ef6003/lib/eslint/eslint.js#L37

File: tests/lib/cli.js [hunk 1]
```diff
@@ -287,6 +287,17 @@ describe("cli", () => {
         });
     });
 
+    describe("when given an async formatter path", () => {
+        it("should execute without any errors", async () => {
+            const formatterPath = getFixturePath("async-formatter.js");
```
Comment (mdjermanovic) [inline tests/lib/cli.js:292]:
```suggestion
            const formatterPath = getFixturePath("formatters", "async-formatter.js");
```

Can we move this file to `fixtures/formatters/`, where the other formatters are?

**Comment by mdjermanovic** [comment]
> Some help is needed, I'd like to modify the `Formatter` type definition, but I just couldn't find where it is from, the `./load-formatter` is missing in codebase.
> 
> https://github.com/eslint/eslint/blob/a1f7ad77e2da00ac7d6daade547fe6bef4ef6003/lib/eslint/eslint.js#L37

Indeed, it's missing. I'd guess that the original plan was to have a separate module for loading formatters and define the type in that file. I searched for `Formatter` and couldn't find the type definition anywhere. It would be very nice to define that type here (in `eslint.js` file, somewhere in the `// Typedefs` section). 

It would be also very nice to update this jsdoc:

https://github.com/eslint/eslint/blob/c9fefd2e40348b3e02b855597707a557dc4991d5/lib/eslint/eslint.js#L627-L634

File: tests/lib/cli.js (hunk 1 above)
Comment (fengzilong) [reply tests/lib/cli.js:292]: I did it before, but other tests failed, not sure why

**Comment by fengzilong** [comment]
Thanks for the tips and suggestion!

File: tests/lib/cli.js (hunk 1 above)
Comment (fengzilong) [reply tests/lib/cli.js]: done @mdjermanovic

**Comment by fengzilong** [comment]
Done, friendly ping~ @mdjermanovic

File: lib/eslint/eslint.js [hunk 2]
```diff
@@ -34,7 +34,12 @@ const { version } = require("../../package.json");
 /** @typedef {import("../shared/types").LintMessage} LintMessage */
 /** @typedef {import("../shared/types").Plugin} Plugin */
 /** @typedef {import("../shared/types").Rule} Rule */
-/** @typedef {import("./load-formatter").Formatter} Formatter */
+
+/**
+ * The main formatter object.
+ * @typedef Formatter
+ * @property {function(LintResult[]): string | Promise<string>} [format] format function.
```
Comment (mdjermanovic) [inline lib/eslint/eslint.js:41]:
```suggestion
 * @property {function(LintResult[]): string | Promise<string>} format format function.
```

`[format]` would mean that it is optional.

File: tests/lib/cli.js (hunk 1 above)
Comment (mdjermanovic) [reply tests/lib/cli.js]: It appears that we're using files in `fixtures/formatters/` as source code files in some tests, so whenever we add a fixture formatter to test how the formatters feature works, we have to update some unrelated tests that are linting those files.

I'll fix this in a follow-up.

**Comment by fengzilong** [comment]
Updated

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

File: docs/developer-guide/working-with-custom-formatters.md [hunk 3]
```diff
@@ -11,6 +11,16 @@ module.exports = function(results) {
 };
 ```
 
+Formatter can also be an async function (from ESLint 8.4.0), the following shows a simple example:
```
Comment (nzakas) [inline docs/developer-guide/working-with-custom-formatters.md:14]:
```suggestion
Formatter can also be an async function (from ESLint v8.4.0), the following shows a simple example:
```

**Review by nzakas — APPROVED** [review]
LGTM. Thanks!

**Review by snitin315 — APPROVED** [review]
LGTM, thank you for your contribution.

**Comment by mdjermanovic** [comment]
Thanks for contributing!

**Comment by nzakas** [comment]
Thanks @fengzilong! We've decided to pay you for this contribution. If you can please send an email to contact (at) eslint (dot) org, we can send you the info.

**Comment by fengzilong** [comment]
Sorry I just saw the message. Thanks for your generosity, but it's only a little work for me, so you don't have to pay me. I'm very happy if this would help more people other than me.

ESLint is an awesome project. If there is any chance, I'd like to contribute more to this project. ❤️

**Pull request merged**